lib/Conversion/MooreToCore/MooreToCore.cpp

```diff
@@ -365,6 +365,34 @@ struct AShrOpConversion : public OpConversionPattern<AShrOp> {
   }
 };
 
+struct ProcedureOpConversion : public OpConversionPattern<ProcedureOp> {
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(ProcedureOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    switch (adaptor.getKind()) {
+    case ProcedureKind::AlwaysComb:
+    case ProcedureKind::Always:
+    case ProcedureKind::AlwaysFF:
+    case ProcedureKind::AlwaysLatch:
+    case ProcedureKind::Final:
+      return emitError(op->getLoc(), "Unsupported procedure operation");
+      return failure();
+    case ProcedureKind::Initial:
+      auto procOp = rewriter.replaceOpWithNewOp<llhd::ProcOp>(
+          op,
+          mlir::FunctionType::get(rewriter.getContext(), std::nullopt,
+                                  std::nullopt),
+          uint64_t(0), mlir::ArrayAttr(), mlir::ArrayAttr());
+      SymbolTable(op->getParentOfType<mlir::ModuleOp>()).insert(procOp);
```

Comment:
@fabianschuiki @hailongSun2000 @cepheus69 I have add llhd.procopto the builtin.module symbol table but get another error.
```
circt-opt: /home/mingzhe/文档/tera-circt/circt/llvm/mlir/lib/IR/SymbolTable.cpp:193: StringAttr mlir::SymbolTable::insert(Operation *, Block::iterator): Assertion `symbol->getParentOp() == symbolTableOp && "symbol is already inserted in another op"' failed.
PLEASE submit a bug report to https://github.com/llvm/circt and include the crash backtrace.
Stack dump:
0.      Program arguments: circt-opt alc.mlir --convert-moore-to-core
 #0 0x000064270d6e868d llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /home/mingzhe/文档/tera-circt/circt/llvm/llvm/lib/Support/Unix/Signals.inc:723:11
 #1 0x000064270d6e8b7b PrintStackTraceSignalHandler(void*) /home/mingzhe/文档/tera-circt/circt/llvm/llvm/lib/Support/Unix/Signals.inc:798:1
 #2 0x000064270d6e6be6 llvm::sys::RunSignalHandlers() /home/mingzhe/文档/tera-circt/circt/llvm/llvm/lib/Support/Signals.cpp:105:5
 #3 0x000064270d6e9335 SignalHandler(int) /home/mingzhe/文档/tera-circt/circt/llvm/llvm/lib/Support/Unix/Signals.inc:413:1
 #4 0x00007726bac42520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #5 0x00007726bac969fc __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #6 0x00007726bac969fc __pthread_kill_internal ./nptl/pthread_kill.c:78:10
 #7 0x00007726bac969fc pthread_kill ./nptl/pthread_kill.c:89:10
 #8 0x00007726bac42476 gsignal ./signal/../sysdeps/posix/raise.c:27:6
 #9 0x00007726bac287f3 abort ./stdlib/abort.c:81:7
#10 0x00007726bac2871b _nl_load_domain ./intl/loadmsgcat.c:1177:9
#11 0x00007726bac39e96 (/lib/x86_64-linux-gnu/libc.so.6+0x39e96)
#12 0x000064270f9b0ea1 mlir::SymbolTable::insert(mlir::Operation*, llvm::ilist_iterator<llvm::ilist_detail::node_options<mlir::Operation, true, false, void, false>, false, false>) /home/mingzhe/文档/tera-circt/circt/llvm/mlir/lib/IR/SymbolTable.cpp:197:35
#13 0x000064270ec7dfa5 (anonymous namespace)::ProcedureOpConversion::matchAndRewrite(circt::moore::ProcedureOp, circt::moore::ProcedureOpAdaptor, mlir::ConversionPatternRewriter&) const /home/mingzhe/文档/tera-circt/circt/lib/Conversion/MooreToCore/MooreToCore.cpp:390:58
#14 0x000064270ec937b2 mlir::OpConversionPattern<circt::moore::ProcedureOp>::matchAndRewrite(mlir::Operation*, llvm::ArrayRef<mlir::Value>, mlir::ConversionPatternRewriter&) const /home/mingzhe/文档/tera-circt/circt/llvm/mlir/include/mlir/Transforms/DialectConversion.h:539:12
#15 0x000064271076eb08 mlir::ConversionPattern::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&) const /home/mingzhe/文档/tera-circt/circt/llvm/mlir/lib/Transforms/Utils/DialectConversion.cpp:1842:10
#16 0x000064271080e02e mlir::PatternApplicator::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&, llvm::function_ref<bool (mlir::Pattern const&)>, llvm::function_ref<void (mlir::Pattern const&)>, llvm::function_ref<mlir::LogicalResult (mlir::Pattern const&)>)::$_0::operator()() const /home/mingzhe/文档/tera-circt/circt/llvm/mlir/lib/Rewrite/PatternApplicator.cpp:212:31
#17 0x000064271080de85 void llvm::function_ref<void ()>::callback_fn<mlir::PatternApplicator::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&, llvm::function_ref<bool (mlir::Pattern const&)>, llvm::function_ref<void (mlir::Pattern const&)>, llvm::function_ref<mlir::LogicalResult (mlir::Pattern const&)>)::$_0>(long) /home/mingzhe/文档/tera-circt/circt/llvm/llvm/include/llvm/ADT/STLFunctionalExtras.h:45:5
#18 0x000064270d5ef859 llvm::function_ref<void ()>::operator()() const /home/mingzhe/文档/tera-circt/circt/llvm/llvm/include/llvm/ADT/STLFunctionalExtras.h:68:5
#19 0x000064271080f885 void mlir::MLIRContext::executeAction<mlir::ApplyPatternAction, mlir::Pattern const&>(llvm::function_ref<void ()>, llvm::ArrayRef<mlir::IRUnit>, mlir::Pattern const&) /home/mingzhe/文档/tera-circt/circt/llvm/mlir/include/mlir/IR/MLIRContext.h:276:3
#20 0x000064271080c927 mlir::PatternApplicator::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&, llvm::function_ref<bool (mlir::Pattern const&)>, llvm::function_ref<void (mlir::Pattern const&)>, llvm::function_ref<mlir::LogicalResult (mlir::Pattern const&)>) /home/mingzhe/文档/tera-circt/circt/llvm/mlir/lib/Rewrite/PatternApplicator.cpp:233:9
#21 0x0000642710777154 (anonymous namespace)::OperationLegalizer::legalizeWithPattern(mlir::Operation*, mlir::ConversionPatternRewriter&) /home/mingzhe/文档/tera-circt/circt/llvm/mlir/lib/Transforms/Utils/DialectConversion.cpp:2150:21
#22 0x000064271076f200 (anonymous namespace)::OperationLegalizer::legalize(mlir::Operation*, mlir::ConversionPatternRewriter&) /home/mingzhe/文档/tera-circt/circt/llvm/mlir/lib/Transforms/Utils/DialectConversion.cpp:2042:17
#23 0x000064271076eb73 mlir::OperationConverter::convert(mlir::ConversionPatternRewriter&, mlir::Operation*) /home/mingzhe/文档/tera-circt/circt/llvm/mlir/lib/Transforms/Utils/DialectConversion.cpp:2573:26
#24 0x000064271076f51f mlir::OperationConverter::convertOperations(llvm::ArrayRef<mlir::Operation*>) /home/mingzhe/文档/tera-circt/circt/llvm/mlir/lib/Transforms/Utils/DialectConversion.cpp:2625:16
#25 0x00006427107742bc mlir::applyFullConversion(llvm::ArrayRef<mlir::Operation*>, mlir::ConversionTarget const&, mlir::FrozenRewritePatternSet const&, mlir::ConversionConfig) /home/mingzhe/文档/tera-circt/circt/llvm/mlir/lib/Transforms/Utils/DialectConversion.cpp:3632:22
#26 0x000064271077435d mlir::applyFullConversion(mlir::Operation*, mlir::ConversionTarget const&, mlir::FrozenRewritePatternSet const&, mlir::ConversionConfig) /home/mingzhe/文档/tera-circt/circt/llvm/mlir/lib/Transforms/Utils/DialectConversion.cpp:3638:10
#27 0x000064270ec55a5d (anonymous namespace)::MooreToCorePass::runOnOperation() /home/mingzhe/文档/tera-circt/circt/lib/Conversion/MooreToCore/MooreToCore.cpp:536:14
#28 0x0000642710e253eb mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int)::$_1::operator()() const /home/mingzhe/文档/tera-circt/circt/llvm/mlir/lib/Pass/Pass.cpp:0:17
#29 0x0000642710e25385 void llvm::function_ref<void ()>::callback_fn<mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int)::$_1>(long) /home/mingzhe/文档/tera-circt/circt/llvm/llvm/include/llvm/ADT/STLFunctionalExtras.h:45:5
#30 0x000064270d5ef859 llvm::function_ref<void ()>::operator()() const /home/mingzhe/文档/tera-circt/circt/llvm/llvm/include/llvm/ADT/STLFunctionalExtras.h:68:5
#31 0x0000642710e28195 void mlir::MLIRContext::executeAction<mlir::PassExecutionAction, mlir::Pass&>(llvm::function_ref<void ()>, llvm::ArrayRef<mlir::IRUnit>, mlir::Pass&) /home/mingzhe/文档/tera-circt/circt/llvm/mlir/include/mlir/IR/MLIRContext.h:276:3
#32 0x0000642710e20b43 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) /home/mingzhe/文档/tera-circt/circt/llvm/mlir/lib/Pass/Pass.cpp:533:17
#33 0x0000642710e210c4 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) /home/mingzhe/文档/tera-circt/circt/llvm/mlir/lib/Pass/Pass.cpp:593:16
#34 0x0000642710e22b08 mlir::PassManager::runPasses(mlir::Operation*, mlir::AnalysisManager) /home/mingzhe/文档/tera-circt/circt/llvm/mlir/lib/Pass/Pass.cpp:904:10
#35 0x0000642710e22a32 mlir::PassManager::run(mlir::Operation*) /home/mingzhe/文档/tera-circt/circt/llvm/mlir/lib/Pass/Pass.cpp:884:60
#36 0x00006427100e7412 performActions(llvm::raw_ostream&, std::shared_ptr<llvm::SourceMgr> const&, mlir::MLIRContext*, mlir::MlirOptMainConfig const&) /home/mingzhe/文档/tera-circt/circt/llvm/mlir/lib/Tools/mlir-opt/MlirOptMain.cpp:409:17
#37 0x00006427100e7048 processBuffer(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::MlirOptMainConfig const&, mlir::DialectRegistry&, llvm::ThreadPoolInterface*) /home/mingzhe/文档/tera-circt/circt/llvm/mlir/lib/Tools/mlir-opt/MlirOptMain.cpp:474:12
#38 0x00006427100e6e2c mlir::MlirOptMain(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::DialectRegistry&, mlir::MlirOptMainConfig const&)::$_0::operator()(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&) const /home/mingzhe/文档/tera-circt/circt/llvm/mlir/lib/Tools/mlir-opt/MlirOptMain.cpp:549:12
#39 0x00006427100e6dc6 mlir::LogicalResult llvm::function_ref<mlir::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>::callback_fn<mlir::MlirOptMain(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::DialectRegistry&, mlir::MlirOptMainConfig const&)::$_0>(long, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&) /home/mingzhe/文档/tera-circt/circt/llvm/llvm/include/llvm/ADT/STLFunctionalExtras.h:45:12
#40 0x00006427101361d2 llvm::function_ref<mlir::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>::operator()(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&) const /home/mingzhe/文档/tera-circt/circt/llvm/llvm/include/llvm/ADT/STLFunctionalExtras.h:68:12
#41 0x00006427101357bc mlir::splitAndProcessBuffer(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::function_ref<mlir::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>, llvm::raw_ostream&, llvm::StringRef, llvm::StringRef) /home/mingzhe/文档/tera-circt/circt/llvm/mlir/lib/Support/ToolUtilities.cpp:28:12
#42 0x00006427100e36f5 mlir::MlirOptMain(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::DialectRegistry&, mlir::MlirOptMainConfig const&) /home/mingzhe/文档/tera-circt/circt/llvm/mlir/lib/Tools/mlir-opt/MlirOptMain.cpp:552:10
#43 0x00006427100e3a59 mlir::MlirOptMain(int, char**, llvm::StringRef, llvm::StringRef, mlir::DialectRegistry&) /home/mingzhe/文档/tera-circt/circt/llvm/mlir/lib/Tools/mlir-opt/MlirOptMain.cpp:590:14
#44 0x00006427100e3c28 mlir::MlirOptMain(int, char**, llvm::StringRef, mlir::DialectRegistry&) /home/mingzhe/文档/tera-circt/circt/llvm/mlir/lib/Tools/mlir-opt/MlirOptMain.cpp:606:10
#45 0x000064270d57ba13 main /home/mingzhe/文档/tera-circt/circt/tools/circt-opt/circt-opt.cpp:76:23
#46 0x00007726bac29d90 __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
#47 0x00007726bac29e40 call_init ./csu/../csu/libc-start.c:128:20
#48 0x00007726bac29e40 __libc_start_main ./csu/../csu/libc-start.c:379:5
#49 0x000064270d57b825 _start (/home/mingzhe/文档/tera-circt/circt/build/bin/circt-opt+0x309d825)
```